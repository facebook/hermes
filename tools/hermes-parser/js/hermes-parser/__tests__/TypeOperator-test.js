/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 */

import {
  printForSnapshotESTree,
  parseForSnapshotESTree,
  printForSnapshotBabel,
  parseForSnapshotBabel,
} from '../__test_utils__/parse';

describe('TypeOperator', () => {
  describe('renders', () => {
    describe('Basic', () => {
      const code = `
type T1 = renders Foo;
type T2 = renders* Foo;
type T3 = renders? Foo;
      `;

      test('ESTree', async () => {
        expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
        expect(await printForSnapshotESTree(code)).toBe(code.trim());
      });

      test('Babel', async () => {
        expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
        expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(`
          "type T1 = any;
          type T2 = any;
          type T3 = any;"
        `);
      });
    });
    describe('Union', () => {
      const code = `
type T1 = renders (Foo | Bar);
type T2 = renders* (Foo | Bar);
type T3 = renders? (Foo | Bar);
      `;

      test('ESTree', async () => {
        expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
        expect(await printForSnapshotESTree(code)).toBe(code.trim());
      });

      test('Babel', async () => {
        expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
        expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(`
          "type T1 = any;
          type T2 = any;
          type T3 = any;"
        `);
      });
    });
    describe('Nested Union', () => {
      const code = `
        type T = renders (Foo | Bar) | null;
      `;

      test('ESTree', async () => {
        expect(await parseForSnapshotESTree(code)).toMatchSnapshot();
        expect(await printForSnapshotESTree(code)).toBe(code.trim());
      });

      test('Babel', async () => {
        expect(await parseForSnapshotBabel(code)).toMatchSnapshot();
        expect(await printForSnapshotBabel(code)).toMatchInlineSnapshot(
          `"type T = any | null;"`,
        );
      });
    });
  });
});
