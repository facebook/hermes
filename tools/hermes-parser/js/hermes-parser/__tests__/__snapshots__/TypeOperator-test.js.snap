// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeOperator renders Basic Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T1",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T2",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T3",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`TypeOperator renders Basic ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T1",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "operator": "renders",
        "type": "TypeOperator",
        "typeAnnotation": {
          "id": {
            "name": "Foo",
            "optional": false,
            "type": "Identifier",
            "typeAnnotation": null,
          },
          "type": "GenericTypeAnnotation",
          "typeParameters": null,
        },
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T2",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "operator": "renders*",
        "type": "TypeOperator",
        "typeAnnotation": {
          "id": {
            "name": "Foo",
            "optional": false,
            "type": "Identifier",
            "typeAnnotation": null,
          },
          "type": "GenericTypeAnnotation",
          "typeParameters": null,
        },
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T3",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "operator": "renders?",
        "type": "TypeOperator",
        "typeAnnotation": {
          "id": {
            "name": "Foo",
            "optional": false,
            "type": "Identifier",
            "typeAnnotation": null,
          },
          "type": "GenericTypeAnnotation",
          "typeParameters": null,
        },
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`TypeOperator renders Nested Union Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T",
        "type": "Identifier",
      },
      "right": {
        "type": "UnionTypeAnnotation",
        "types": [
          {
            "type": "AnyTypeAnnotation",
          },
          {
            "type": "NullLiteralTypeAnnotation",
          },
        ],
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`TypeOperator renders Nested Union ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "type": "UnionTypeAnnotation",
        "types": [
          {
            "operator": "renders",
            "type": "TypeOperator",
            "typeAnnotation": {
              "type": "UnionTypeAnnotation",
              "types": [
                {
                  "id": {
                    "name": "Foo",
                    "optional": false,
                    "type": "Identifier",
                    "typeAnnotation": null,
                  },
                  "type": "GenericTypeAnnotation",
                  "typeParameters": null,
                },
                {
                  "id": {
                    "name": "Bar",
                    "optional": false,
                    "type": "Identifier",
                    "typeAnnotation": null,
                  },
                  "type": "GenericTypeAnnotation",
                  "typeParameters": null,
                },
              ],
            },
          },
          {
            "type": "NullLiteralTypeAnnotation",
          },
        ],
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`TypeOperator renders Union Babel 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T1",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T2",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T3",
        "type": "Identifier",
      },
      "right": {
        "type": "AnyTypeAnnotation",
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;

exports[`TypeOperator renders Union ESTree 1`] = `
{
  "body": [
    {
      "id": {
        "name": "T1",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "operator": "renders",
        "type": "TypeOperator",
        "typeAnnotation": {
          "type": "UnionTypeAnnotation",
          "types": [
            {
              "id": {
                "name": "Foo",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "type": "GenericTypeAnnotation",
              "typeParameters": null,
            },
            {
              "id": {
                "name": "Bar",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "type": "GenericTypeAnnotation",
              "typeParameters": null,
            },
          ],
        },
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T2",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "operator": "renders*",
        "type": "TypeOperator",
        "typeAnnotation": {
          "type": "UnionTypeAnnotation",
          "types": [
            {
              "id": {
                "name": "Foo",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "type": "GenericTypeAnnotation",
              "typeParameters": null,
            },
            {
              "id": {
                "name": "Bar",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "type": "GenericTypeAnnotation",
              "typeParameters": null,
            },
          ],
        },
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
    {
      "id": {
        "name": "T3",
        "optional": false,
        "type": "Identifier",
        "typeAnnotation": null,
      },
      "right": {
        "operator": "renders?",
        "type": "TypeOperator",
        "typeAnnotation": {
          "type": "UnionTypeAnnotation",
          "types": [
            {
              "id": {
                "name": "Foo",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "type": "GenericTypeAnnotation",
              "typeParameters": null,
            },
            {
              "id": {
                "name": "Bar",
                "optional": false,
                "type": "Identifier",
                "typeAnnotation": null,
              },
              "type": "GenericTypeAnnotation",
              "typeParameters": null,
            },
          ],
        },
      },
      "type": "TypeAlias",
      "typeParameters": null,
    },
  ],
  "type": "Program",
}
`;
